### PR TITLE
Update Azure default image

### PR DESCRIPTION
**Pull request by cblmemo**
This PR updates the default Azure Image to `ubuntu-hpc` 20230803, an image with the latest NVIDIA driver(535.54.03) and CUDA version (12.2).

Reference: https://github.com/Azure/azhpc-images/releases/tag/ubuntu-hpc-20230803

For context, seems like all `ubuntu-hpc` image is V2 so I still use the original image for V1 instances.

Tested (run the relevant ones):

- [ ] Code formatting: `bash format.sh`
- [ ] Any manual or new tests for this PR (please specify below)
- [x] All smoke tests: `pytest tests/test_smoke.py` 
     - Except for several TPU tests, see #2469 
- [ ] Relevant individual smoke tests: `pytest tests/test_smoke.py::test_fill_in_the_name` 
- [ ] Backward compatibility tests: `bash tests/backward_comaptibility_tests.sh`

### CODE REVIEW

File: sky/clouds/azure.py [hunk 1]
```diff
@@ -136,35 +136,40 @@ def get_default_instance_type(
     def _get_image_config(self, gen_version, instance_type):
         # az vm image list \
         #  --publisher microsoft-dsvm --all --output table
-        # nvidia-driver: 495.29.05, cuda: 11.5
-
-        # The latest image 2022.09.14/2022.08.11/22.06.10/22.05.11/
-        # 22.04.27/22.04.05 has even older nvidia driver 470.57.02,
-        # cuda: 11.4
+        # nvidia-driver: 535.54.03, cuda: 12.2
+        # see: https://github.com/Azure/azhpc-images/releases
```
Comment (infwinston) [inline sky/clouds/azure.py:140]:
```suggestion
        # see: https://github.com/Azure/azhpc-images/releases/tag/ubuntu-hpc-20230803
```
let's link to specific tag. otherwise it will change in the future

File: sky/clouds/azure.py [hunk 2]
```diff
@@ -136,35 +136,40 @@ def get_default_instance_type(
...
-            'image_version': '21.11.04'
+            'image_offer': 'ubuntu-hpc',
+            'image_sku': '2204',
+            'image_version': '22.04.2023080201'
         }
 
-        # ubuntu-2004 v21.10.21 and v21.11.04 do not work on K80
-        # due to an NVIDIA driver issue.
+        # ubuntu-2004 v221.08.30, one image we used to use for K80 before
+        # we change default image to ubuntu-hpc.
```
Comment (infwinston) [inline sky/clouds/azure.py:151]:
```suggestion
        # ubuntu-2004 v21.08.30, K80 requires image with old NVIDIA driver version
```

File: sky/clouds/azure.py [hunk 3]
```diff
@@ -136,35 +136,40 @@ def get_default_instance_type(
...
-            image_config['image_version'] = '20.04.2021120101'
+                image_config = {
+                    'image_publisher': 'microsoft-dsvm',
+                    'image_offer': 'ubuntu-2004',
+                    'image_sku': '2004-gen2',
+                    'image_version': '21.08.30'
+                }
+
+        # ubuntu-2004 v21.11.04, one image we used to use for V1 HyperV
+        # before we change default image to ubuntu-hpc.
```
Comment (infwinston) [inline sky/clouds/azure.py:164]: what does this mean?

**Review by infwinston — COMMENTED** [review]
thanks @cblmemo !

File: sky/clouds/azure.py (hunk 3 above)
Comment (cblmemo) [reply sky/clouds/azure.py:163]: The `gen_version` is reference to HyperV generation IIUC? https://github.com/skypilot-org/skypilot/blob/8125ebf19980461f14676edf2c6fe126122c0c7d/sky/clouds/service_catalog/data_fetchers/fetch_azure.py#L218

**Comment by cblmemo** [comment]
> thanks @cblmemo! as this PR upgrades CUDA version to 12. we want to make sure it works for common GPUs. Can we test T4?

I just realized there is not a lot of tests in our default smoke test... Will run with `--azure` flag again

**Comment by infwinston** [comment]
We could just test it using the TGI example. with T4 + 3b model? if this works we can merge it.

File: sky/authentication.py [hunk 4]
```diff
@@ -267,10 +268,31 @@ def setup_gcp_authentication(config: Dict[str, Any]) -> Dict[str, Any]:
     return _replace_ssh_info_in_config(config, public_key)
 
 
+# Cloud init script is encoded in base64 in azure, so we need to decode it
+# and replace the ssh user and public key content, then encode it back.
```
Comment (Michaelvll) [inline sky/authentication.py:272]: Do we really need to encode the script with base64?

File: sky/clouds/azure.py [hunk 5]
```diff
@@ -249,13 +253,15 @@ def make_deploy_resources_variables(
         # This script will modify /etc/ssh/sshd_config and add a bash script
         # into .bashrc. The bash script will restart sshd if it has not been
         # restarted, identified by a file /tmp/__restarted is existing.
+        # Also, add default user to docker group.
         # pylint: disable=line-too-long
         cloud_init_setup_commands = base64.b64encode(
```
Comment (Michaelvll) [inline sky/clouds/azure.py:258]: Same as above, do we really need to encode the cloud init to base64? Is it possible to use a plain text with some careful yaml syntax, e.g. using `: |` or `: |-` to avoid unexpected new line and indents?

**Review by Michaelvll — COMMENTED** [review]
Thanks for updating this @cblmemo! Left some comments about the cloud init.

File: sky/authentication.py (hunk 4 above)
Comment (cblmemo) [reply sky/authentication.py:272]: Yes... Azure only accepts base64 encoded in the `customData` field. See https://github.com/skypilot-org/skypilot/pull/1910#discussion_r1262930925 for more information

Comment (cblmemo) [reply sky/authentication.py:272]: Let me add a comment for this

File: sky/authentication.py [hunk 6]
```diff
@@ -267,10 +268,33 @@ def setup_gcp_authentication(config: Dict[str, Any]) -> Dict[str, Any]:
     return _replace_ssh_info_in_config(config, public_key)
 
 
+# In Azure, cloud-init script must be encoded in base64. See
+# https://learn.microsoft.com/en-us/azure/virtual-machines/custom-data
+# for more information. Here we decode it and replace the ssh user
+# and public key content, then encode it back.
+def _replace_ssh_info_in_cloud_init(config: Dict[str, Any],
```
Comment (Michaelvll) [inline sky/authentication.py:275]: This seems only be used by Azure, should we have it inline in `setup_azure_authentication`?

File: sky/clouds/azure.py (hunk 3 above)
Comment (Michaelvll) [reply sky/clouds/azure.py:163]: The comment does not read. Could you rephrase the comment? Aslo, does the new `ubuntu-hpc` work for gen `V1`?

**Review by Michaelvll — COMMENTED** [review]
Thanks for the fix @cblmemo! Left several comments.

File: sky/clouds/azure.py (hunk 3 above)
Comment (cblmemo) [reply sky/clouds/azure.py:163]: No, it only supports V2. To double-check, you can use:
```python
import json
import subprocess
from concurrent.futures import ThreadPoolExecutor

images_output = subprocess.check_output(
    "az vm image list --publisher microsoft-dsvm --offer ubuntu-hpc --all", shell=True
).decode("utf-8")

images = json.loads(images_output)
print(f'Found {len(images)} images')


def check_image(image):
    output = (
        subprocess.check_output(
            (
                "az vm image show "
                f'--publisher {image["publisher"]} '
                f'--offer {image["offer"]} '
                f'--sku {image["sku"]} '
                f'--version {image["version"]} '
                '--query "hyperVGeneration" -o tsv'
            ),
            shell=True,
        )
        .decode("utf-8")
        .strip()
    )
    print(output)


with ThreadPoolExecutor(max_workers=10) as executor:
    list(executor.map(check_image, images))
```

File: sky/clouds/azure.py [hunk 7]
```diff
@@ -136,35 +136,39 @@ def get_default_instance_type(
...
+        # the latest ubuntu-hpc:2204 image.
         image_config = {
             'image_publisher': 'microsoft-dsvm',
-            'image_offer': 'ubuntu-2004',
-            'image_sku': '2004-gen2',
-            'image_version': '21.11.04'
+            'image_offer': 'ubuntu-hpc',
+            'image_sku': '2204',
+            'image_version': '22.04.2023080201'
         }
```
Comment (Michaelvll) [inline sky/clouds/azure.py:148]: Considering we are adding more different images for Azure, we should have the `image.csv` design for Asure as well. Can we add a TODO here for the refactoring in the future?

File: sky/clouds/azure.py [hunk 8]
```diff
@@ -136,35 +136,39 @@ def get_default_instance_type(
...
+                image_config = {
+                    'image_publisher': 'microsoft-dsvm',
+                    'image_offer': 'ubuntu-2004',
+                    'image_sku': '2004-gen2',
+                    'image_version': '21.08.30'
+                }
+
+        # ubuntu-2004 v21.11.04, the previous image we used in the past for
+        # V1 HyperV instance before we change default image to ubuntu-hpc.
         if gen_version == 'V1':
```
Comment (Michaelvll) [inline sky/clouds/azure.py:164]: Just wondering, what instances are labeled `V1`? Can we have some examples / explanations in the comment here?

**Review by Michaelvll — APPROVED** [review]
Thanks for updating the image for Azure instances @cblmemo! cc'ing @romilbhardwaj to have a test on this PR for #2751.

**Comment by cblmemo** [comment]
Fixed - Please let me know when you finish testing @romilbhardwaj and I'll merge this 🫡

**Comment by Michaelvll** [comment]
The latest PR seems working with the following `nvidia-smi` output:

```
Thu Nov 16 23:26:09 2023       
+---------------------------------------------------------------------------------------+
| NVIDIA-SMI 535.54.03              Driver Version: 535.54.03    CUDA Version: 12.2     |
|-----------------------------------------+----------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |         Memory-Usage | GPU-Util  Compute M. |
|                                         |                      |               MIG M. |
|=========================================+======================+======================|
|   0  NVIDIA A100 80GB PCIe          Off | 00000001:00:00.0 Off |                    0 |
| N/A   28C    P0              41W / 300W |      4MiB / 81920MiB |      0%      Default |
|                                         |                      |             Disabled |
```

Please feel free to merge it @cblmemo. : )

**Comment by datainvestor** [comment]
Is this only available on nightly? Because I keep getting error than the cuda driver is less than 11.8 when running Azure VM with A100 GPU
